### ISSUE TITLE
Add methods to calculate a pixel or a position from tile coordinates in tile matrix sets

### ISSUE THREAD

**Issue by navispatial** [navibyte/geospatial#158]
Related to the feature request in #157.

Add new methods to `TileMatrixSet` / `GeoTileMatrixSet` and implementations.

Drafting method signatures for `GeoTileMatrixSet:

```dart
enum TileAligment { 
  center, 
  topLeft, 
  // ...
}

Projected tileToWorld(Scalable2i tile, 
      {TileAligment alignment = TileAligment.center});
Scalable2i tileToPixel(Scalable2i tile, 
      {TileAligment alignment = TileAligment.center});
Geographic tileToPosition(Scalable2i tile, 
      {TileAligment alignment = TileAligment.center});
```

**Comment by navispatial** [comment]
Other considerations how to define `aligment`, see the [Aligment](https://api.flutter.dev/flutter/painting/Alignment-class.html) class in Flutter as defined by it:

> A point within a rectangle.
> 
> Alignment(0.0, 0.0) represents the center of the rectangle. The distance from -1.0 to +1.0 is the distance from one side of the rectangle to the other side of the rectangle. Therefore, 2.0 units horizontally (or vertically) is equivalent to the width  (or height) of the rectangle.
> 
> Alignment(-1.0, -1.0) represents the top left of the rectangle.
> 
> Alignment(1.0, 1.0) represents the bottom right of the rectangle.
> 
> Alignment(0.0, 3.0) represents a point that is horizontally centered with respect to the rectangle and vertically below the bottom of the rectangle by the height of the rectangle.
> 
> Alignment(0.0, -0.5) represents a point that is horizontally centered with respect to the rectangle and vertically half way  between the top edge and the center.
> 
> Alignment(x, y) in a rectangle with height h and width w describes the point (x * w/2 + w/2, y * h/2 + h/2) in the coordinate system of the rectangle.

This would be very handy to be used on planned methods instead of an enum. However using it would mean dependency on Flutter libraries that's not possible as `geobase` is kept as a pure Dart library.

Related also [Rect](https://api.flutter.dev/flutter/dart-ui/Rect-class.html), [Offset](https://api.flutter.dev/flutter/dart-ui/Offset-class.html) and [Size](https://api.flutter.dev/flutter/dart-ui/Size-class.html) in `dart:ui`.

**Comment by navispatial** [comment]
And there is also [FractionalOffset](https://api.flutter.dev/flutter/painting/FractionalOffset-class.html) in Flutter too with interesting history descriptions..

**Comment by navispatial** [comment]
Not to forget [AlignmentDirectional](https://api.flutter.dev/flutter/painting/AlignmentDirectional-class.html)

**Comment by navispatial** [comment]
`Aligned` class:

```dart
/// An aligned point within a geospatial box or tile.
///
/// Inspired by the
/// [Alignment](https://api.flutter.dev/flutter/painting/Alignment-class.html)
/// and `AlignmentDirectional` classes defined by the Flutter SDK.
///
/// This class is named `Aligned` to avoid name collisions with `Alignment` and
/// `Align` classes defined by the Flutter SDK.
@immutable
class Aligned {
  /// The horizontal distance fraction.
  ///
  /// The value `-1.0` represents the west side edge of the box.
  ///
  /// The value `0.0` represents the center horizontally.
  ///
  /// The value `1.0` represents the east side edge of the box.
  final double x;

  /// The vertical distance fraction.
  ///
  /// The value `-1.0` represents the south side edge of the box.
  ///
  /// The value `0.0` represents the center vertically.
  ///
  /// The value `1.0` represents the north side edge of the box.
  final double y;

  /// An aligned point within a geospatial box or tile.
  const Aligned({required this.x, required this.y});

  /// The south west corner, with `x: -1.0, y: -1.0`.
  static const southWest = Aligned(x: -1.0, y: -1.0);

  /// The center point along the south side edge, with `x: 0.0, y: -1.0`.
  static const southCenter = Aligned(x: 0.0, y: -1.0);

  /// The south east corner, with `x: 1.0, y: -1.0`.
  static const southEast = Aligned(x: 1.0, y: -1.0);

  /// The center point along the west side edge, with `x: -1.0, y: 0.0`.
  static const centerWest = Aligned(x: -1.0, y: 0.0);

  /// The center point, with `x: 0.0, y: 0.0`.
  static const center = Aligned(x: 0.0, y: 0.0);

  /// The center point along the east side edge, with `x: 1.0, y: 0.0`.
  static const centerEast = Aligned(x: 1.0, y: 0.0);

  /// The north west corner, with `x: -1.0, y: 1.0`.
  static const northWest = Aligned(x: -1.0, y: 1.0);

  /// The center point along the north side edge, with `x: 0.0, y: 1.0`.
  static const northCenter = Aligned(x: 0.0, y: 1.0);

  /// The north east corner, with `x: 1.0, y: 1.0`.
  static const northEast = Aligned(x: 1.0, y: 1.0);

  // ...
}
```

New methdods on tile matrix set classes too, definition in `TileMatrixSet`:

```dart
 /// Transforms a fractional point defined by [align] inside [tile] to world
  /// coordinates.
  ///
  /// By default with `Aligned.center` the world coorinates at the center of a
  /// tile are returned.
  ///
  /// Coordinate value ranges:
  /// - tile x (int): `0 .. matrixWidth(zoom) - 1`
  /// - tile y (int): `0 .. matrixHeight(zoom) - 1`
  /// - world x (double): `0.0 .. mapWidth(0)`
  /// - world y (double): `0.0 .. mapHeight(0)`
  Projected tileToWorld(Scalable2i tile, {Aligned align = Aligned.center});

  /// Transforms a fractional point defined by [align] inside [tile] to a
  /// position.
  ///
  /// By default with `Aligned.center` the position at the center of a tile is
  /// returned.
  ///
  /// Coordinate value ranges:
  /// - tile x (int): `0 .. matrixWidth(zoom) - 1`
  /// - tile y (int): `0 .. matrixHeight(zoom) - 1`
  Position tileToPosition(Scalable2i tile, {Aligned align = Aligned.center});

  /// Transforms a fractional point defined by [align] inside [tile] to pixel
  /// coordinates.
  ///
  /// By default with `Aligned.center` the pixel at the center of a tile is
  /// returned.
  ///
  /// When [insideTile] is true, it's guaranteed that the target pixel is inside
  /// the source tile.
  ///
  /// Coordinate value ranges:
  /// - tile x (int): `0 .. matrixWidth(zoom) - 1`
  /// - tile y (int): `0 .. matrixHeight(zoom) - 1`
  /// - pixel x (int): `0 .. mapWidth(zoom) - 1`
  /// - pixel y (int): `0 .. mapHeight(zoom) - 1`
  Scalable2i tileToPixel(
    Scalable2i tile, {
    Aligned align = Aligned.center,
    bool insideTile = false,
  });
```

Overridden signature in `GeoTileMatrixSet`:

```dart
@override
  Geographic tileToPosition(Scalable2i tile, {Aligned align = Aligned.center});
```

New sample code in `geobase_example.dart`:

```dart
  // aligned points (world, pixel and position coordinates) inside tile or edges
  print(quad.tileToWorld(tile, align: Aligned.northWest));
  print(quad.tileToPixel(tile, align: Aligned.center));
  print(quad.tileToPosition(tile, align: Aligned.center));
  print(quad.tileToPosition(tile, align: Aligned.southEast));

  // get zoomed tile at the center of a source tile
  final centerOfTile2 = quad.tileToWorld(tile, align: Aligned.center);
  final tile7 = quad.worldToTile(centerOfTile2, zoom: 7);
  print('tile at zoom 2: $tile => center of tile: $centerOfTile2 '
      '=> tile at zoom 7: $tile7');
```

**Comment by navispatial** [comment]
Implemented in [geobase 0.4.0-dev.0](https://pub.dev/packages/geobase/versions/0.4.0-dev.0)

**Comment by navispatial** [comment]
Note: `Aligned.center` like in sample above `centerOfTile2 = quad.tileToWorld(tile, align: Aligned.center)` returns a center point of a tile in projected coordinate space.

For example if Web Mercator tile matrix set is used, a center point of a projected tile is not the same point as a center point calculated from geographical bounding box derived from corner points of a tile.
